internal/mesh/internal/controllers/implicitdestinations/controller_test.go

```diff
@@ -0,0 +1,1544 @@
+			}
+		} else {
+			status = &pbresource.Status{
+				Conditions: []*pbresource.Condition{{
+					Type:    catalog.StatusConditionBoundIdentities,
+					State:   pbresource.Condition_STATE_TRUE,
+					Message: "",
+				}},
+			}
+		}
+		r := rtest.Resource(pbcatalog.ServiceType, n).
+			WithData(suite.T(), &pbcatalog.Service{
+				Workloads: &pbcatalog.WorkloadSelector{
+					Names: workloads,
+				},
+				Ports: []*pbcatalog.ServicePort{
+					{TargetPort: "mesh", Protocol: pbcatalog.Protocol_PROTOCOL_MESH},
+					{TargetPort: "grpc", Protocol: pbcatalog.Protocol_PROTOCOL_GRPC},
+				},
+			}).
+			WithTenancy(tenancy).
+			WithStatus(catalog.EndpointsStatusKey, status).
+			Write(suite.T(), suite.client)
+
+		var statusUpdate = func() *pbresource.Resource { return r }
+		if !withWIDs {
+			statusUpdate = func() *pbresource.Resource {
+				ctx := suite.client.Context(suite.T())
+
+				status := &pbresource.Status{
+					ObservedGeneration: r.Generation,
+					Conditions: []*pbresource.Condition{{
+						Type:    catalog.StatusConditionBoundIdentities,
+						State:   pbresource.Condition_STATE_TRUE,
+						Message: strings.Join(ids, ","),
+					}},
+				}
+				resp, err := suite.client.WriteStatus(ctx, &pbresource.WriteStatusRequest{
+					Id:     r.Id,
+					Key:    catalog.EndpointsStatusKey,
+					Status: status,
+				})
+				require.NoError(suite.T(), err)
+				return resp.Resource
+			}
+		}
+
+		sf := &serviceFixture{
+			Resource:     r,
+			StatusUpdate: statusUpdate,
+		}
+
+		rs = append(rs, sf)
+	}
+	return rs
+}
+
+func (suite *controllerSuite) createComputedRoutes(svc *pbresource.Resource, decResList ...any) *types.DecodedComputedRoutes {
+	resList := make([]any, 0, len(decResList)+1)
+	resList = append(resList,
+		resourcetest.MustDecode[*pbcatalog.Service](suite.T(), svc),
+	)
+	resList = append(resList, decResList...)
+	crID := resource.ReplaceType(pbmesh.ComputedRoutesType, svc.Id)
+	cr := routestest.ReconcileComputedRoutes(suite.T(), suite.client, crID, resList...)
+	require.NotNil(suite.T(), cr)
+	return cr
+}
+
+func (suite *controllerSuite) TestReconcile_CIDCreate_NoReferencingTrafficPermissionsExist() {
+	suite.runTestCaseWithTenancies(func(tenancy *pbresource.Tenancy) {
+		wi := rtest.Resource(pbauth.WorkloadIdentityType, "wi1").
+			WithTenancy(tenancy).
+			Write(suite.T(), suite.client)
+
+		id := rtest.Resource(pbmesh.ComputedImplicitDestinationsType, wi.Id.Name).
+			WithTenancy(tenancy).
+			ID()
+
+		suite.reconcileOnce(id)
+
+		// Ensure that the CID was created
+		cid := suite.client.RequireResourceExists(suite.T(), id)
+		suite.requireCID(cid, &pbmesh.ComputedImplicitDestinations{
+			Destinations:    []*pbmesh.ImplicitDestination{},
+			BoundReferences: nil,
+		})
+	})
+}
+
+func (suite *controllerSuite) TestReconcile_CIDCreate_ReferencingResourcesExist() {
+	suite.runTestCaseWithTenancies(func(tenancy *pbresource.Tenancy) {
+		suite.createTrafficPermissions([]string{"d-wi1-s-wi2"}, []string{"wi2"}, tenancy)
```

Comment:
I find that I'm needing to jump around the code a ton to understand what these helper functions are doing. I don't have any ideas for making this more explicit without adding thousands of lines to have the same coverage though. So, I think we should keep this as is.